File: src/main/java/engine/rendering/instances/postprocessing/gammacorrection/GammaCorrection.java
package engine.rendering.instances.postprocessing.gammacorrection;

import engine.rendering.RenderOutputData;
import engine.rendering.abstracted.postprocessing.AbstractPostProcessor;
import engine.utils.libraryBindings.opengl.shaders.RenderState;
import engine.utils.libraryBindings.opengl.shaders.UniformFloatProperty;
import engine.utils.libraryBindings.opengl.shaders.UniformTextureProperty;
import engine.utils.libraryBindings.opengl.textures.ITexture;

public class GammaCorrection extends AbstractPostProcessor {

  private static final String FRAG_FILE = "/shaders/postprocessing/gammacorrection/GammaCorrectionFragment.glsl";
  private float factor;

  public GammaCorrection(float factor) throws Exception {
    super(FRAG_FILE);
    this.factor = factor;
    initialize();
  }

  public GammaCorrection() throws Exception {
    this(2.2f); // Default value
  }

  private void initialize() {
    getShadersProgram().addPerRenderUniform(new UniformTextureProperty<RenderOutputData>("textureSampler", 0) {
      @Override
      public ITexture getUniformValue(RenderState<RenderOutputData> state) {
        return state.getInstance().colour;
      }
    });

    getShadersProgram().addPerRenderUniform(new UniformFloatProperty<RenderOutputData>("factor") {
      @Override
      public float getUniformValue(RenderState<RenderOutputData> state) {
        return factor;
      }
    });
  }
}

File: src/main/java/engine/rendering/instances/postprocessing/gammacorrection/GammaCorrection.kt
package engine.rendering.instances.postprocessing.gammacorrection

import engine.rendering.RenderOutputData
import engine.rendering.abstracted.postprocessing.AbstractPostProcessor
import engine.utils.libraryBindings.opengl.shaders.RenderState
import engine.utils.libraryBindings.opengl.shaders.UniformFloatProperty
import engine.utils.libraryBindings.opengl.shaders.UniformTextureProperty
import engine.utils.libraryBindings.opengl.textures.ITexture

class GammaCorrection(var factor: Float = 2.2f) : AbstractPostProcessor(FRAG_FILE) {

    init {
        shadersProgram.addPerRenderUniform(object : UniformTextureProperty<RenderOutputData>("textureSampler", 0) {
            override fun getUniformValue(state: RenderState<RenderOutputData>): ITexture {
                return state.instance.colour
            }
        })
        shadersProgram.addPerRenderUniform(object : UniformFloatProperty<RenderOutputData>("factor") {
            override fun getUniformValue(state: RenderState<RenderOutputData>): Float {
                return factor
            }
        })
    }

    companion object {
        private const val FRAG_FILE = "/shaders/postprocessing/gammacorrection/GammaCorrectionFragment.glsl"
    }

}
